text/3451-additional-float-types.md

```diff
@@ -0,0 +1,139 @@
+
+```rust
+#[cfg(target_has_f128)]
+fn get_f128() -> f128 { 1.0f128 }
+
+#[cfg(target_has_f16)]
+fn get_f16() -> f16 { 1.0f16 }
+```
+
+All operators, constants and math functions defined for `f32` and `f64` in core, are also defined for `f16` and `f128`, and guarded by respective conditional guards.
+
+`f80` type is defined in `core::arch::{x86, x86_64}`. `doubledouble` type is defined in `core::arch::{powerpc, powerpc64}`. `bf16` type is defined in `core::arch::{arm, aarch64, x86, x86_64}`. They do not have literal representation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `f16` type
+
+`f16` consists of 1 bit of sign, 5 bits of exponent, 10 bits of mantissa.
+
+The following `From` and `TryFrom` traits are implemented for conversion between `f16` and other types:
+
+```rust
+impl From<f16> for f32 { /* ... */ }
+impl From<f16> for f64 { /* ... */ }
+impl From<bool> for f16 { /* ... */ }
+impl From<u8> for f16 { /* ... */ }
+impl From<i8> for f16 { /* ... */ }
+```
+
+`f16` will generate `half` type in LLVM IR.
+
+## `f128` type
+
+`f128` consists of 1 bit of sign, 15 bits of exponent, 112 bits of mantissa.
+
+`f128` is available for on targets having (1) hardware instructions or software emulation for 128-bit float type; (2) backend support for `f128` type on the target; (3) essential target features enabled (if any).
+
+The list of targets supporting `f128` type may change over time. Initially, it includes `powerpc64le-*`.
+
+The following traits are also implemented for conversion between `f128` and other types:
+
+```rust
+impl From<f16> for f128 { /* ... */ }
+impl From<f32> for f128 { /* ... */ }
+impl From<f64> for f128 { /* ... */ }
+impl From<bool> for f128 { /* ... */ }
+impl From<u8> for f128 { /* ... */ }
+impl From<i8> for f128 { /* ... */ }
+impl From<u16> for f128 { /* ... */ }
+impl From<i16> for f128 { /* ... */ }
+impl From<u32> for f128 { /* ... */ }
+impl From<i32> for f128 { /* ... */ }
+impl From<u64> for f128 { /* ... */ }
+impl From<i64> for f128 { /* ... */ }
+```
+
+`f128` will generate `fp128` type in LLVM IR.
+
+
+`std::simd` defines new vector types with `f16` or `f128` element: `f16x2` `f16x4` `f16x8` `f16x16` `f16x32` `f128x2` `f128x4`.
+
+For `doubledouble` type, conversion intrinsics are available under `core::arch::{powerpc, powerpc64}`. For `f80` type, conversion intrinsics are available under `core::arch::{x86, x86_64}`.
+
+## Architectures specific types
+
+As for non-standard types, `f80` generates `x86_fp80`, `doubledouble` generates `ppc_fp128`, `bf16` generates `bfloat`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Unlike f32 and f64, although there are platform independent implementation of supplementary intrinsics on these types, not every target support the two types natively, with regards to the ABI. Adding them will be a challenge for handling different cases.
```

Comment:
For riscv without hardware float support there is a defined soft-float ABI. There is not for f16/bf16. Same for x86_64. Many other architectures likely don't have a defined soft-float abi for f128 either. And as I understand it AArch64 doesn't have a soft-float abi at all as Neon support is mandatory and floats are even allowed inside the kernel unlike eg x86_64 where floats are disabled in the kernel to avoid having to save them on syscalls.